### ISSUE TITLE
Potential path traversal when using filename from Zip archive

### ISSUE THREAD

**Issue by h3xstream** [find-sec-bugs/find-sec-bugs#514]
## Description

The class [ZipEntry ](https://docs.oracle.com/javase/8/docs/api/java/util/zip/ZipEntry.html)describe one file inside a Zip archive. [`getName()`](https://docs.oracle.com/javase/8/docs/api/java/util/zip/ZipEntry.html#getName--) return the file name from this file. It could contains a malicious string such as "../../../".

Reference: https://blog.ripstech.com/2019/hidden-flaws-of-archives-java/

## Vulnerable code

```java
Enumeration entries = zip.entries();
while (entries.hasMoreElements()) {
    ZipEntry zipEntry = entries.nextElement();
    File file = new File(outputDir, zipEntry.getName())
[...]
```

API to cover : 

- [`java.util.zip.ZipEntry.getName()`](https://docs.oracle.com/javase/8/docs/api/java/util/zip/ZipEntry.html#getName--)
- `org.apache.commons.compress.archivers.ar.ArArchiveEntry.getName()`
- `org.apache.commons.compress.archivers.arj.ArjArchiveEntry.getName()`
- `org.apache.commons.compress.archivers.cpio.CpioArchiveEntry.getName()`
- `org.apache.commons.compress.archivers.dump.DumpArchiveEntry.getName()`
- `org.apache.commons.compress.archivers.jar.JarArchiveEntry.getName()`
- `org.apache.commons.compress.archivers.sevenz.SevenZArchiveEntry.getName()`
- `org.apache.commons.compress.archivers.tar.TarArchiveEntry.getName()`
- `org.apache.commons.compress.archivers.zip.ZipArchiveEntry.getName()`
- [`org.apache.commons.compress.archivers.ArchiveEntry.getName()`](https://commons.apache.org/proper/commons-compress/apidocs/org/apache/commons/compress/archivers/ArchiveEntry.html)

## Similar detector

If anybody is interested in the implementation of such rule. It is very similar to the rule to detect `FileItem.getName()` from multipart upload.

https://github.com/find-sec-bugs/find-sec-bugs/blob/f86e2f2ba90ad39191bc9e88597628adbabd36f7/findsecbugs-plugin/src/main/java/com/h3xstream/findsecbugs/file/FileUploadFilenameDetector.java#L31-L51

**Comment by h3xstream** [comment]
Bonus  : Could be implemented in the same enhancements.. 
zip4j is a popular library. I think that the equivalent concept would be [AbstractFileHeader.getFileName()](https://github.com/srikanth-lingala/zip4j/blob/master/src/main/java/net/lingala/zip4j/model/AbstractFileHeader.java#L110)

**Comment by thiyagu-7** [comment]
Can I take this up? :)

**Comment by h3xstream** [comment]
@thiyagu-7 Yes

**Comment by thiyagu-7** [comment]
I modified one of the existing test classes adding 
 
    Enumeration<? extends ZipEntry> entries = zip.entries();
    while (entries.hasMoreElements()) {
        ZipEntry e = entries.nextElement();
        File f = new File("C:/Code/", e.getName());
    }

This raised `PATH_TRAVERSAL_IN` bug as implemented by [PathTraversalDetector](https://github.com/find-sec-bugs/find-sec-bugs/blob/2340b90bc3993a595011e1cce20982533799cc0d/findsecbugs-plugin/src/main/java/com/h3xstream/findsecbugs/file/PathTraversalDetector.java).

    New Bug Instance: [1]
      message=SECPTI: This API (java/io/File.<init>(Ljava/lang/String;Ljava/lang/String;)V) reads a file whose location might be specified by user input
      bugType=PATH_TRAVERSAL_IN  priority=Medium  category=S
      class=testcode.FileUploadCommon  method=handleFile  line=33
      sources=[[java/util/zip/ZipEntry.getName()Ljava/lang/String;]]

Would this then also cover the ZipEntry cases too? 🤔

**Comment by h3xstream** [comment]
@thiyagu-7 Yes the fact that there is both a PATH_TRAVERSAL_IN  and the new ZIP_ENTRY_NAME? being raised is perfect.

Both have a different purpose.  `ZipEntry.getName()` is the risky **source**. `File()` is the method from which malicious value should not reach unfiltered (**sink**).

**Comment by h3xstream** [comment]
This issue is up for grab.

**Comment by akshaycbor** [comment]
Can I take it?

**Comment by akshaycbor** [comment]
How do I detect that the value read in getName reaches the File() (sink) unfiltered?

**Comment by h3xstream** [comment]
@akshaycbor Yes

**Comment by akshaycbor** [comment]
@h3xstream Maybe a silly question but how do I get a single test to run. Whenever I try to run a single test using `mvn -Dtest=FileUploadFilenameDetectorTest test` it fails

**Comment by h3xstream** [comment]
@akshaycbor Running the test individually also failed on my side. I did not find a solution running a single test with Maven.

I have mostly use Maven to run the complete suite and use my IDE IntelliJ to run the test individually.

I will investigated this later.

**Comment by Ayan07** [comment]
@h3xstream I 'm interested to work on this issue.
I see this issue is currently unassigned.
Can I take this up?
